### PR TITLE
feature/improved-termination-notification

### PR DESCRIPTION
**Pull request by johantjellden**
Send events to notification-service on termination events
* Think about should we include the terminationdate updated event
Relates to https://github.com/HedvigInsurance/notification-service/pull/496

### CODE REVIEW

File: src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/dto/EventRequest.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+data class ContractCreatedEvent(
+  val contractId: String,
+  val owningMemberId: String,
+  val startDate: LocalDate?,
+  val signSource: String? = null
+) : EventRequest()
+
+data class ContractTerminatedEvent(
+  val contractId: String,
+  val owningMemberId: String,
```
Comment (palmenhq) [inline src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/dto/EventRequest.kt:37]: idk if this is too late to change but isn't `holderMemberId` what we use everywhere else?

**Review by palmenhq — APPROVED** [review]
Looks fine! Just one thought

File: src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/NotificationServiceClient.java [hunk 2]
```diff
@@ -34,4 +32,8 @@
 
   @PostMapping("/_/events/contractRenewalQueued")
```
Comment (AlexSJ1990) [inline src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/NotificationServiceClient.java:33]: shall we add the deprecated annotation to these

File: src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/NotificationServiceClient.java [hunk 3]
```diff
@@ -34,4 +32,8 @@
...
   void sendContractRenewalQueued(@RequestBody ContractRenewalQueuedEvent data, @RequestHeader("Request-Id") String eventId);
+
+  @PostMapping("/_/event")
+  void sendContractTerminatedEvent(@NotNull ContractTerminatedEvent contractTerminatedEvent, @RequestHeader("Request-Id") String identifier);
```
Comment (AlexSJ1990) [inline src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/NotificationServiceClient.java:37]: maybe `void sendEvent(NotNull EventRequest eventRequest, RequestHeader("Request-Id") String identifier)`

**Review by AlexSJ1990 — CHANGES_REQUESTED** [review]

File: src/main/java/com/hedvig/contract/customerio/EventListener.kt [hunk 4]
```diff
@@ -170,38 +172,42 @@ class EventListener(
...
   @EventHandler
   fun on(
     event: ContractTerminationDateChanged,
-    @Timestamp timestamp: Instant
+    @Timestamp timestamp: Instant,
+    eventMessage: DomainEventMessage<Any>
   ) {
     onlyRunIfEventIsNewerThanComponent(timestamp, event) {
-      sendCancellation(event, timestamp, event.newTerminationDate, event.contractId)
+      sendCancellation(event, timestamp, event.newTerminationDate, event.contractId, eventMessage.identifier)
```
Comment (vonElfvin) [inline src/main/java/com/hedvig/contract/customerio/EventListener.kt:201]: I wouldn't send an email if we change the date I think? Maybe update some attribute though? 🤔

File: src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/NotificationServiceImpl.kt [hunk 5]
```diff
@@ -107,4 +108,24 @@ class NotificationServiceImpl(val notificationServiceClient: NotificationService
...
+      notificationServiceClient.sendContractTerminatedEvent(
+        ContractTerminatedEvent(
+          contractId = contractId.toString(),
+          owningMemberId = memberId,
+          terminationDate = terminationDate
+        ),
+        identifier
+      )
+    } catch (ex: FeignException) {
+      logger.error("Could not update notification-service for event $contractId", ex)
```
Comment (vonElfvin) [inline src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/NotificationServiceImpl.kt:128]:
```suggestion
      logger.error("Could not update notification-service for event $identifier (contractId=$contractId)", ex)
```

File: src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/dto/EventRequest.kt [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+data class StartDateUpdatedEvent(
+  val contractId: String,
+  val owningMemberId: String,
+  val startDate: LocalDate
+) : EventRequest()
+
+enum class ChargeFailedReason {
+  INSUFFICIENT_FUNDS,
+  NOT_CONNECTED_DIRECT_DEBIT
+}
```
Comment (vonElfvin) [inline src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/dto/EventRequest.kt:74]: Do we need this? Or maybe, in the future, this should be in some kind of package?

**Review by vonElfvin — COMMENTED** [review]
Added some comments

File: src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/dto/EventRequest.kt (hunk 1 above)
Comment (johantjellden) [reply src/main/java/com/hedvig/productPricing/service/serviceIntegration/notificationService/dto/EventRequest.kt:37]: Yup to late, these DTO are already in use else where.

File: src/main/java/com/hedvig/contract/customerio/EventListener.kt (hunk 4 above)
Comment (johantjellden) [reply src/main/java/com/hedvig/contract/customerio/EventListener.kt:201]: Why not is it not good to now that we now have a new cancellationdate?

**Review by AlexSJ1990 — APPROVED** [review]

**Pull request merged**